cas-server-support-pac4j/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/AbstractClientAuthenticationHandler.java

```diff
@@ -0,0 +1,121 @@
+import org.jasig.cas.authentication.principal.DefaultPrincipalFactory;
+import org.jasig.cas.authentication.principal.PrincipalFactory;
+import org.jasig.cas.support.pac4j.authentication.principal.ClientCredential;
+import org.pac4j.core.client.Client;
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.context.J2EContext;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.profile.UserProfile;
+import org.springframework.webflow.context.ExternalContextHolder;
+import org.springframework.webflow.context.servlet.ServletExternalContext;
+
+import javax.security.auth.login.FailedLoginException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.validation.constraints.NotNull;
+import java.security.GeneralSecurityException;
+
+/**
+ * Generic handler for delegated authentication: it uses the client credentials to get the user profile
+ * returned by the provider for an authenticated user.
+ *
+ * @author Jerome Leleu
+ * @since 4.1.0
+ */
+@SuppressWarnings("unchecked")
+public abstract class AbstractClientAuthenticationHandler extends AbstractPreAndPostProcessingAuthenticationHandler {
+
+    /** Factory to create the principal type. **/
+    @NotNull
+    protected PrincipalFactory principalFactory = new DefaultPrincipalFactory();
+
+    /**
+     * The clients for authentication.
+     */
+    @NotNull
+    private final Clients clients;
+
+    /**
+     * Define the clients.
+     *
+     * @param theClients The clients for authentication
+     */
+    public AbstractClientAuthenticationHandler(final Clients theClients) {
+        this.clients = theClients;
+    }
+
+    @Override
+    public boolean supports(final Credential credential) {
+        return credential != null && ClientCredential.class.isAssignableFrom(credential.getClass());
+    }
+
+    @Override
+    protected HandlerResult doAuthentication(final Credential credential) throws GeneralSecurityException, PreventedException {
+        final ClientCredential clientCredentials = (ClientCredential) credential;
+        logger.debug("clientCredentials : {}", clientCredentials);
+
+        final String clientName = clientCredentials.getCredentials().getClientName();
+        logger.debug("clientName : {}", clientName);
+
+        // get client
+        final Client<Credentials, UserProfile> client = this.clients.findClient(clientName);
+        logger.debug("client : {}", client);
+
+        // web context
+        final ServletExternalContext servletExternalContext = (ServletExternalContext) ExternalContextHolder.getExternalContext();
+        final HttpServletRequest request = (HttpServletRequest) servletExternalContext.getNativeRequest();
+        final HttpServletResponse response = (HttpServletResponse) servletExternalContext.getNativeResponse();
+        final WebContext webContext = new J2EContext(request, response);
```

Comment:
Ah, you're right! Never mind. Didnt realize createResult is doing that. In that case, I dont think adding request and response is helpful. If needed, one could always override doAuthentication(). 

Thanks for your patience with me.